File: javascript/signing.js
var https = require('https');
var crypto = require('crypto');
var pp = require('prettyjson');

var access_key = process.env.ES_ACCESS_KEY;
var secret_key = process.env.ES_SECRET_KEY;

var http_method = 'GET'
var api_version = '2012-02-29';
var sign_path = '/api/enstratus/'+api_version+'/geography/Cloud';
var ua = 'enstratus.js';

var hmac = crypto.createHmac('sha256', secret_key);
var timestamp = Date.now();

var parts = [access_key, http_method, sign_path, timestamp, ua];
var to_sign = parts.join(':');

hmac.update(to_sign);
var b64auth = hmac.digest('base64');

var es_headers = { 'x-esauth-access': access_key,
		'x-esauth-timestamp': timestamp, 
		'x-esauth-signature': b64auth,
		'x-es-details': 'basic', 
		'accept': 'application/json', 
		'user-agent': ua };


https.get({ host: 'api.enstratus.com', path: sign_path, headers: es_headers }, function(res) {
	console.log(res.headers);

	var resp = '';

	res.on('data', function(d) {
		resp += d;
	});

	res.on('end', function () {
		console.log(pp.render(JSON.parse(resp)));	
	});
}).on('error', function(e) {
	console.error(e);
});

File: ruby/signing.rb
require 'excon'
require 'openssl'
require 'json'
require 'base64'
require 'pp'

# normally I'd do anything other than access_key and secret_key as an options hash
def sign_request(access_key, secret_key, ua='enstratus.rb', http_method='GET', path='/geography/Cloud', api_version='2012-02-29')

  timestamp = (Time.now.to_i * 1000)
  sign_path = "/api/enstratus/#{api_version}#{path}"

  parts = []
  parts << access_key << http_method << sign_path << timestamp << ua
  to_sign = parts.join(":")

  digest = OpenSSL::HMAC.digest(OpenSSL::Digest::Digest.new('sha256'), secret_key, to_sign)
  b64auth = Base64.encode64(digest).strip
    
end

access_key = ENV['ES_ACCESS_KEY']
secret_key = ENV['ES_SECRET_KEY']
ua = 'enstratus.rb'
api_version = '2012-02-29'
path = '/geography/Cloud'
http_method='GET'
timestamp = (Time.now.to_i * 1000)

signature = sign_request(access_key, secret_key, ua=ua, http_method=http_method, path=path, api_version=api_version)

connection = Excon.new('https://api.enstratus.com')

headers = {'User-agent' => ua,
           'x-esauth-access' => access_key,
           'x-esauth-timestamp' => timestamp,
           'x-esauth-signature' => signature,
           'x-es-details' => 'basic',
           'Accept' => 'application/json'}

results = connection.get(:path => "/api/enstratus/#{api_version}/geography/Cloud", :headers => headers)

pp results.headers
pp JSON.load(results.body)
